### DISCUSSION TITLE
Configurable navbar

### DISCUSSION THREAD

**Discussion by LuHugo** [withastro/starlight#963] [Feature Requests]
### What version of `starlight` are you using?

0.10.2

### What is your idea?

As the title says, we need a configurable navigation bar menu.
The navbar option can be added to the starlight configuration item. navbar contains the data needed to generate the menu。
For example： 

```js
export default defineConfig({
  integrations: [starlight({
       ...
	navbar: {
                showSearch: false,
                showLocale: false,
		items: [
			{
				label: "Home",
				to: "/",
				position: "left",
				activeBaseRegex: '/',
				type: "link",
			},
			{
				label: "Reference",
				to: "/reference",
				position: "left",
				activeBaseRegex: '/reference',
				type: "dropdown",
				items: [MenuItem]
			}
		]
	}
  })]
});
```

### Why is this feature necessary?

Can organize website navigation more freely

### Do you have examples of this feature in other projects?

- [docusaurus navbar](https://docusaurus.io/docs/api/themes/configuration#navbar-items)
- [vuepress navbar](https://vuepress.vuejs.org/theme/default-theme-config.html#navbar-links)

### Participation

- [ ] I am willing to submit a pull request for this issue.

**Comment by delucis** [comment]
Thanks for the issue @LuHugo! We’ll soon release #709, which will make it quite easy to add a custom Astro component to the navbar for use cases like this. I’d love for us to wait and see whether that works well for people or whether we still need a custom configuration-based API.

**Reply by dionysuzx** [reply]
hello @delucis thanks for the feature impl. perhaps describing my issue (related to this) may help in getting support for how to achieve what i want:
- lets say i add a custom link to the navbar by overriding the social icons component, this link is just a link to the "Docs"
- this is because finding the "docs" is difficult on mobile if most people are used to it being in the navbar instead of a CTA

however, the overridden component disappears because i think there is some logic that hides it once the mobile breakpoint is hit. is there a way to customize that behavior as well?

the end goal is basically the ability to have some links in the navbar that persist on mobile view.

**Reply by dionysuzx** [reply]
i'd like to add some more context for others. i do know that i can remove the "splash" page from the frontmatter, which in some sense removes the need for links in the navbar because the user is already obviously "within" the docs. you can see an example here on the astro docs https://docs.astro.build/en/getting-started/ (attached a picture below).

In this case the "astro icon" and the "docs icon" on the upper left corner are separate links. so i believe (haven't tested it) you can just add a link in this fashion, and it will persist even in the mobile view. this could be a way to achieve what was described above.

anyways, not a make it or break it feature for me, so curious on others' thoughts. my general take is that it would be nice to give the flexibility if we're able to figure out a design for the dropdown that still feels intuitive / non-cluttered.

<img width="2560" alt="image" src="https://github.com/withastro/starlight/assets/13951458/d1f5907f-0a54-452f-beb7-d669df585c65">

**Comment by levino** [comment]
>I’d love for us to wait and see whether that works well for people or whether we still need a custom configuration-based API.

It does not work well for me. I want to use starlight on a page with other, non-starlight pages (using additional content collections, so they cannot be simply markdown files.). I am able to add links from the navigation menu to these pages by overriding the `Header` component. However this feels way too heavy to just add a couple of additional navigation entries.

But the real problem is, that I fail to use the `@astrojs/starlight/index.astro` as a template for my other pages. Of course I want to use the same navigation header for my other pages so I do not have to implement it twice. However the component needs specific `props` and I am failing to produce these.

Please again have a look at how docusaurus has [solved](https://docusaurus.io/docs/api/themes/configuration#navbar-items) this very conveniently for the users. Basically the idea should be that you have pages AND docs AND a blog etc. So instead of "taking over everything" starlight should be more of an additional feature / part of the page. Right now its usage basically cripples my site in the sense that I can only use markdown. Maybe this could be solved by refactoring the `index.astro` file to use some kind of `Template.astro` and exporting this component for the users.

**Reply by delucis** [reply]
Hey @levino, sounds like you’re describing #1175! Hopefully we can ship that soon.

**Reply by levino** [reply]
If anyone is interested, I started some [work](https://github.com/levino/levinkeller.de/tree/a12b8ee41d891e31b6918a2cbe294bb549bd7163/shipyard) on a package with working title "shipyard". The idea is to separate the layout and design from the content collection specifications as I [proposed](https://github.com/withastro/starlight/pull/1175#issuecomment-1875241916) in #1175. Unfortunately I think that unless the tight coupling is removed in starlight, starlight is kinda broken by design. At least if one wants to build something more complex than just a documentation website. "Shipyard" is supposed to eventually become a collection of building blocks to build new websites with astro fast, with or without a docs section.

I am aware that the amount of work necessary to build this package on my own is significant and I probably will fail without help. So I thought I would leave this here. If someone wants to give me a hand, I am more than willing to accept their help.

**Reply by levino** [reply]
@delucis I appreciate the feature for [custom pages using the starlight layout](https://starlight.astro.build/guides/pages/#custom-pages), however this does not solve the issue with the non-configurable navigation bar.

**Reply by delucis** [reply]
Sharing this plugin here which adds support for nav bar links that may be helpful: https://starlight-utils.pages.dev/utilities/nav-links/

(Also shared in a comment lower on this page)

**Reply by levino** [reply]
The plugin is a nice attempt, but [fundamentally broken](https://github.com/lorenzolewis/starlight-utils/issues/72). Largely because the design of starlight is just not flexible enough, I fear. As I said elsewhere, the unnecessary tight coupling of the "astro templating" and the "docs functionality" is in my opinion the main reason for this. Docusaurus solved this elegantly by making the docs feature an optional plugin and by that leaves the page much more configurable for the users. I would assume is also helps during development to not constrain oneself on the docs functionality in order to build a more general usable site generator.

**Reply by delucis** [reply]
[Astro](https://astro.build/) itself is an fully flexible static site generator with Starlight being the “docs plugin”, so I think the parallel is already present? You can use a regular `.astro` page template in `src/pages/` and generate pages that look however you want.

**Reply by levino** [reply]
Thank you for taking the time to come back to me on this. I am aware of the capabilities of astro and I think mostly of starlight. However docusaurus has simply made some very clever design choices which makes it not only suitable for documentation websites but also standalone pages, blogs or a combination of all these features within one page. The only missing pieces from docusaurus are:
- server side rendering for per route authentication
- support of content collections for bulk page generation
- 
But is a mean tool to build content-driven websites real fast. What does such a software need?
- A simple and consistent design / layout which supports usual pages, docs pages and blog pages
- Auto generated navigation if desired (for blog and docs)
- Configurable theming and navigation

Starlight provides the configurable theming and the auto generated navigation. However it always brings "docs" which a lot of people might not need. They simply want a blog. Why not use starlight for this too? Currently they have to jump hoops to add a blog functionality to starlight and the addon (at least last time I checked) had to do some heavy plumbing in order to get things working.

It would just help to separate concerns like in docusaurus where there is first the simple basic layout and page generation and then docs etc. come as plugins. Docs and blog come with the boilerplate code during scaffolding, so no problem with user experience. I really think that docusaurus made a lot of the api decisions very nicely (maybe by luck, maybe by skill, I cannot say) and that starlight would benefit from adopting their api. It could then become a general purpose page builder and make docusaurus obsolete.

**Reply by delucis** [reply]
Aha, thanks for the details. So IIUC you’re saying that the basic templating concepts are good (page layout, standardised nav bar, routing, dark/light theme, search, etc.) but you’d like more flexibility in the _type_ of pages available to make using it for other purposes easier while still maintaining a consistent layout, components etc.?

**Reply by levino** [reply]
I find starlight is pretty far in this regard. One can make arbitrary pages and [reuse the starlight layout](https://starlight.astro.build/guides/pages/#using-starlights-design-in-custom-pages). The only missing puzzle piece is a configurable "global" navigation (a navigation which shows on all pages and behaves nicely on mobile, where it potentially needs to integrate with the docs navigation for lack of space).

However I had a look of how this functionality is implemented, and my analyses is that it was done with much pain simply because of the architecture choices that were made for starlight, most importantly the tight coupling of the "docs functionality" and the "layout and design". Instead of "decorating the docs pages" one has to "undecorate the non-docs pages" which feels suboptimal to me.

**Reply by levino** [reply]
Also maybe for understanding the background a bit: I am using docusaurus to build community pages for small clubs (think gardening, sports). They often want to write down some rules, maybe have a list of people responsible for certain areas etc. and I use "docs" for that. Then they want to record their meeting notes, give updates on certain topics, etc. and I use "blog" for that. And then they want some general pages like "about", "imprint" etc. Maybe they also have multiple sections in the club so each section needs their own docs or something but the blog is for the whole club.

Also I wanted to integrate some more content based functionality in one of these sites, which is a garden planning tool where one can use one content collection to specify a group of plants for a bed and then see when do to what. This stuff simply does not work with docusaurus, so I switched to astro but then I had to bother with all the design stuff and the navigation generation for the docs etc. I saw starlight and thought: Cool, it is all there already, but unfortunately some crucial things (like having a navigation menu for the about and the imprint) does not work. This is why I more or less rewrote starlight from scratch https://github.com/levino/levinkeller.de/tree/main/shipyard. It is a very poor attempt at achieving what you guys have done, but in my opinion it fixes a big mistake that was made when starlight was initially created. If this mistake was fixed, it would have far greater potential to be a page builder for any website (which would all look pretty much the same, but who cares? Done is better than perfect, hm?)

**Reply by delucis** [reply]
Thank you! I appreciate all the context and detailed feedback. 

Starlight's origin is from building out our docs template to be easier to use and adapt for more people, but for sure that led to some decisions that are restrictive to other uses (and there are also a few spots — like in any software project probably — where we cut a corner or two to get things working for the intended use case that also make customisations harder than they could be).

I hope we'll be able to address some of these pain points in the future.

**Comment by sagargurtu** [comment]
This is a much needed basic feature in Starlight. Many other static site generators offer a way to add navbar menu items with dropdowns. Some examples:
- [Vitepress](https://vitepress.dev/reference/default-theme-nav#navigation-links)
- [Docusaurus](https://docusaurus.io/docs/api/themes/configuration#navbar-items)

**Comment by vauxia** [comment]
I got a basic site up and running quickly, but this issue has been bogging me down me down for hours. I agree this is a very basic feature, and it should not be difficult for new users.

As far as I can tell, there is no way to add navbar items. You can, however, override an _adjacent_ component (theme selector, title, etc) and shoehorn in some links before/after the otherwise-unrelated component.

From a software organization/architecture standpoint, this is yucky and indirect. 

From an onboarding perspective, it creates a lot of pain and confusion for a new user task that should be very simple. ( For instance, I am now trying to troubleshoot an error in the Search  component because I overrode the SiteTitle component, because I wanted to add some general links in a universally expected place on the page..? )

**Reply by delucis** [reply]
Thanks for the feedback!

**Reply by jarodtaylor** [reply]
This is the situation I ran into. I was able to create multiple sidebars using the link provided by @delucis in [#687 comment](https://github.com/withastro/starlight/pull/687#issuecomment-2043417045). Works wonderfully using the component overrides!

My problem now is I'd like navigation links in the Header to go to these different sections of the site. This allows us to break out our documentation into sections like (API, Apps & Services, Theming, etc). Otherwise, the sidebar will be massive. 

I tried to override the [Header.astro](packages/starlight/components/Header.astro) component, but as described above, it slots my Navigation Links outside of the rest of the header elements. Similar to @vauxia I tried doing this in the other Header components (e.g. SiteTitle.astro) but even that has its issues.

**Comment by jbergstroem** [comment]
I also found the complexity of adding top level items with a sidebar that shows the state of said page instead of the top level sidebar much higher than expected. There is [a plugin for that](https://starlight-utils.pages.dev/) but it will not work once you start overriding components. I don't mind adapting my overridden components from starlight, but then to merge this with plugin-specific logic you ultimately end up just rebuilding it for yourself.

**Comment by andrezero** [comment]
I arrived here becauise I was able to add the custom pages and routes from a colleciton, but couldn't add them to the sidebar. 

While I appreciate tha autogenerated docs / guides/ ... structure, the particular dccs I am putting together are 90% generated from metadata, retrieived in JSON from various sources. 

It would be awesome to be able to just use Astro collections to feed items into the sidebar

⚠️  this is NOT a workaround `import { getCollection } from 'astro:content';` is only available in `.astro` code AFAIK

```js
import { defineConfig } from 'astro/config';
import starlight from '@astrojs/starlight';

import { getCollection } from 'astro:content';
const metas = await getCollection('metas');

export default defineConfig({
    experimental: { contentLayer: true },
    integrations: [
        starlight({
            title: '[Sandbox]',
            sidebar: [
                {
                    label: 'Metas',
                    items: metas.map(meta => meta.id),
                },
            ],
        }),
    ],
});
```
⚠️  this is NOT a workaround `import { getCollection } from 'astro:content';` is only available in `.astro` code AFAIK

Since sidebar is currently defined at configuration time, I have no idea how to work around this. 

--- 

EDIT: I also tried to statically generate a JSON file from the same source that feeds the collection, but Starlight performs a check for each item, requiring it to be `a valid entry slug in the docs content collection.`.

Of course, this is a nice feature, preventing errors (and promoting good practices in docs authoring).

```
AstroUserError
An error occurred.
The slug "components/parametric-viz" specified in the Starlight sidebar config does not exist.
````

But it feels a bit harsh

:question:  Would it possible to somehow opt-out of this check? 

Maybe providing `path` instead of `slug` could denote that we want to link to something we KNOW exists elsewhere, i.e., not from the `docs` collection. 

```js
sidebar: [
    {
        label: 'Components',
        items: [{
            path: 'components/parametric-viz' // path instead of slug
        }],
    },
]
```

💡 This would allow to link to any `/foo/bar` in our `docs.domain.com/`. For all we care, it could exist because it was generated from a collection, or it is being redirected somewhere in Cloudlare 😅 

Food for thought. 
Keep up the greast work 🏅

**Comment by sahilarora535** [comment]
For those still stuck with this, [`Tauri`'s website](https://tauri.app/) does it well by overriding components. See: https://github.com/tauri-apps/tauri-docs/tree/v2/src/components/overrides

Although I would still +1 on this. Having an ability to configure the navbar easily should be a part of a static site generator by default. If this can be possible, this would make lives super easy. All the mature static site generators, `Jekyll`, `Hugo`, `Docusaurus`, etc. have this feature. 

The [Starlight Blog Plugin](https://starlight-blog-docs.vercel.app/) adds a `Blog` link in the header, but that's not visible on `mobile` devices, and feels hacky. Something like what's done in [`Tauri`'s website](https://tauri.app/) would be amazing.

Here is how I think this should be:

| Desktop | Mobile |
| --- | --- | 
| <img width="1916" height="994" alt="image" src="https://github.com/user-attachments/assets/32314eb5-c05a-4aa8-a924-d05504511a51" /> | <img width="429" height="879" alt="image" src="https://github.com/user-attachments/assets/646dc12f-6a71-4bac-9da2-ad4937a82c87" /> |

CC: @delucis @HiDeoo

**Reply by HiDeoo** [reply]
> The [Starlight Blog Plugin](https://starlight-blog-docs.vercel.app/) adds a `Blog` link in the header, but that's not visible on `mobile` devices, and feels hacky.

Only commenting on the blog plugin comment as I'm the author: as you've mentioned, a link is added to the header by default on large viewports. On smaller viewports, the link is added to the mobile menu sidebar for non-blog pages. Such behavior can also be configured using the [`navigation` option](https://starlight-blog-docs.vercel.app/configuration/#navigation) to add the link after the site title or logo, before the theme switcher (default), or to not show it at all so users can add it manually wherever they want.

**Comment by mrfoureyed** [comment]
Man... been hot min since this was originally opened, curious if this is still something that will be implemented. Since the roll out of custom plugins still nothing has solved this issue and it hasn't been added to the codebase.

**Reply by levino** [reply]
Maybe shipyard.levinkeller.de is something for you. I am very open for feature requests.
